backend/src/main/java/de/amos/apachepulsarui/service/ClusterService.java

```diff
@@ -52,4 +60,37 @@ private ClusterDto enrichWithBrokerData(ClusterDto cluster) {
         }
     }
 
+    /**
+     * Aggregates the topology elements by iterating in a breadth-first search manner top-down:
+     * all clusters
+     * -> tenants (all of cluster)
+     * -> namespaces (all of tenant)
+     * -> topics (all of namespace)
+     *
+     * @return A list of nested topology elements from clusters till namespaces.
+     */
+    private ClusterDto enrichWithTopologyElements(ClusterDto cluster) {
+
+        List<TenantDto> tenantsWithNamespacesAndTopics = tenantService.getAllTenants().stream()
+                // filter all tenants for those that are allowed for our cluster
+                .filter(tenant -> tenant.getTenantInfo()
+                        .getAllowedClusters()
+                        .contains(cluster.getId()))
+                // set all the namespaces of a tenant
+                .peek(tenant -> {
+                    List<NamespaceDto> namespacesOfTenant = namespaceService.getAllOfTenant(tenant).stream()
+                            // for each namespace of a tenant, we fetch all topics
+                            .peek(namespace -> {
+                                List<String> topicsOfNamespace = topicService.getAllNamesByNamespace(namespace.getId());
+                                namespace.setTopics(topicsOfNamespace);
+                            })
```

Comment:
This is already done as part of `namespaceService.getAllOfTenant()`, I think the peek can be removed : )

Reply:
Nice! Removed it.